src/coerce.c

```diff
@@ -18,21 +54,48 @@ double logical_to_real(int x) {
 double integer_to_real(int x) {
   return (x == NA_INTEGER) ? NA_REAL : x;
 }
-SEXP logical_to_char(int x) {
-  if (x == NA_LOGICAL)
-    return NA_STRING;
 
-  return Rf_mkChar(x ? "TRUE" : "FALSE");
+void deprecate_to_char(const char* type) {
+  SEXP fn = PROTECT(Rf_lang3(
+    Rf_install(":::"),
+    Rf_install("purrr"),
+    Rf_install("deprecate_to_char")
+  ));
+
+  SEXP call = PROTECT(Rf_lang2(
+    PROTECT(fn),
+    PROTECT(Rf_mkString("type"))
+  ));
+
+  Rf_eval(call, R_BaseEnv);
+  UNPROTECT(4);
 }
-SEXP integer_to_char(int x) {
+
+SEXP logical_to_char(int x, SEXP from, SEXP to, int i) {
+  if (x == NA_LOGICAL) {
+    return NA_STRING;
+  } else {
+    if (i == 0)
+      deprecate_to_char("logical");
```

Comment:
@lionel- This is a bit of a hack because I want to deprecate aggressively so `deprecate_to_char()` uses `always = TRUE` and then I only call it for the first element. There's some small risk that the deprecation message won't get generated if the transformation is not type stable (e.g. `map_chr(list("x", 1), identity)`) but I think it's worth taking that small risk is return for much simpler code.